File: airbyte-integrations/bases/base-java/src/test/java/io/airbyte/integrations/base/FailureTrackingAirbyteMessageConsumerTest.java
/*
 * Copyright (c) 2022 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.integrations.base;

import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doThrow;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import io.airbyte.protocol.models.v0.AirbyteMessage;
import io.airbyte.protocol.models.v0.AirbyteMessage.Type;
import org.junit.jupiter.api.Test;

class FailureTrackingAirbyteMessageConsumerTest {

  @Test
  void testStartNoFailure() throws Exception {
    final TestConsumer consumer = spy(new TestConsumer());
    consumer.start();
    consumer.close();

    verify(consumer).close(false);
  }

  @Test
  void testStartWithFailure() throws Exception {
    final TestConsumer consumer = spy(new TestConsumer());
    doThrow(new RuntimeException()).when(consumer).startTracked();

    // verify the exception still gets thrown.
    assertThrows(RuntimeException.class, consumer::start);
    consumer.close();

    verify(consumer).close(true);
  }

  @Test
  void testAcceptNoFailure() throws Exception {
    final TestConsumer consumer = spy(new TestConsumer());

    final AirbyteMessage msg = mock(AirbyteMessage.class);
    consumer.accept(msg);
    consumer.close();

    verify(consumer).close(false);
  }

  @Test
  void testAcceptWithFailure() throws Exception {
    final TestConsumer consumer = spy(new TestConsumer());
    final AirbyteMessage msg = mock(AirbyteMessage.class);
    when(msg.getType()).thenReturn(Type.RECORD);
    doThrow(new RuntimeException()).when(consumer).acceptTracked(any());

    // verify the exception still gets thrown.
    assertThrows(RuntimeException.class, () -> consumer.accept(msg));
    consumer.close();

    verify(consumer).close(true);
  }

  static class TestConsumer extends FailureTrackingAirbyteMessageConsumer {

    @Override
    protected void startTracked() {

    }

    @Override
    protected void acceptTracked(final AirbyteMessage s) {

    }

    @Override
    protected void close(final boolean hasFailed) {

    }

  }

}

File: airbyte-cdk/java/airbyte-cdk/core/src/test/kotlin/io/airbyte/cdk/integrations/base/FailureTrackingAirbyteMessageConsumerTest.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.cdk.integrations.base

import io.airbyte.protocol.models.v0.AirbyteMessage
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import org.mockito.kotlin.any
import org.mockito.kotlin.mock

internal class FailureTrackingAirbyteMessageConsumerTest {
    @Test
    @Throws(Exception::class)
    fun testStartNoFailure() {
        val consumer = Mockito.spy(TestConsumer())
        consumer.start()
        consumer.close()

        Mockito.verify(consumer).close(false)
    }

    @Test
    @Throws(Exception::class)
    fun testStartWithFailure() {
        val consumer = Mockito.spy(TestConsumer())
        Mockito.doThrow(RuntimeException()).`when`(consumer).startTracked()

        // verify the exception still gets thrown.
        Assertions.assertThrows(RuntimeException::class.java) { consumer.start() }
        consumer.close()

        Mockito.verify(consumer).close(true)
    }

    @Test
    @Throws(Exception::class)
    fun testAcceptNoFailure() {
        val consumer = Mockito.spy(TestConsumer())

        val msg = Mockito.mock(AirbyteMessage::class.java)
        consumer.accept(msg)
        consumer.close()

        Mockito.verify(consumer).close(false)
    }

    @Test
    @Throws(Exception::class)
    fun testAcceptWithFailure() {
        val consumer = Mockito.spy(TestConsumer())
        val msg: AirbyteMessage = mock()
        Mockito.`when`(msg.type).thenReturn(AirbyteMessage.Type.RECORD)
        Mockito.doThrow(RuntimeException()).`when`(consumer).acceptTracked(any())

        // verify the exception still gets thrown.
        Assertions.assertThrows(RuntimeException::class.java) { consumer.accept(msg) }
        consumer.close()

        Mockito.verify(consumer).close(true)
    }

    internal class TestConsumer : FailureTrackingAirbyteMessageConsumer() {
        public override fun startTracked() {}

        public override fun acceptTracked(s: AirbyteMessage) {}

        public override fun close(hasFailed: Boolean) {}
    }
}
